pyclam/chaoda.py

```diff
@@ -0,0 +1,129 @@
+""" This module provides the CHAODA algorithms implemented on top of CLAM.
+"""
+
+from typing import List, Dict, Set
+
+import numpy as np
+from scipy.special import erf
+
+from pyclam import Graph, Cluster
+
+_NORMALIZATION_MODES = ['linear', 'gaussian', 'sigmoid']
+
+
+def _normalize(scores: np.array, mode: str) -> np.array:
+    if mode == 'linear':
+        min_v, max_v, = float(np.min(scores)), float(np.max(scores))
+        if min_v == max_v:
+            max_v += 1.
+        scores = (scores - min_v) / (max_v - min_v)
+    else:
+        mu: float = float(np.mean(scores))
+        sigma: float = max(float(np.std(scores)), 1e-3)
+
+        if mode == 'gaussian':
+            scores = erf((scores - mu) / (sigma * np.sqrt(2)))
+        elif mode == 'sigmoid':
+            scores = 1 / (1 + np.exp(-(scores - mu) / sigma))
+        else:
+            raise ValueError(f'normalization mode {mode} is undefined. Use one of {_NORMALIZATION_MODES}.')
+
+    return scores.ravel().clip(0, 1)
+
+
+def cluster_cardinality(graph: Graph, normalize: str = 'gaussian') -> np.array:
+    """ Determines outlier scores for points by considering the relative cardinalities of the clusters in the graph.
+
+    Points in clusters with relatively low cardinalities are the outliers.
+
+    :param graph: Graph on which to calculate outlier scores.
+    :param normalize: which normalization mode to use to get scores in a [0, 1] range.
+                      Must be one of 'linear', 'gaussian', or 'sigmoid'.
+    :return: A numpy array of outlier scores for each point in the manifold that the graph belongs to.
+    """
+    if normalize not in _NORMALIZATION_MODES:
+        raise ValueError(f'Normalization method {normalize} is undefined. Must by one of {_NORMALIZATION_MODES}.')
+
+    scores: Dict[int, int] = {p: -cluster.cardinality for cluster in graph.clusters for p in cluster.argpoints}
```

Comment:
These two lines are a bit odd to parse, is the first line intended to filter out duplicate points in the graph?

Reply:
There are no duplicates in the graph. That's part of the Graph Invariant. Scores are supposed to be for points rather than for clusters. there are repeated scores, i.e. values, but not repeated keys, i.e. point indices.